controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
-// returns placement decisions which can be the decisions for only a subset of subscriptions
-//
-func (a *AVRInstance) processSubscriptions() bool {
-	a.log.Info("Process subscriptions", "total", len(a.subscriptionList.Items))
+func (r *ApplicationVolumeReplicationReconciler) getUserPlacementRule(ctx context.Context,
+	avr *rmn.ApplicationVolumeReplication) (*plrv1.PlacementRule, error) {
+	r.Log.Info("Getting User PlacementRule", "placement", avr.Spec.Placement)
 
-	requeue := false
+	if avr.Spec.Placement == nil || avr.Spec.Placement.PlacementRef == nil {
+		return nil, fmt.Errorf("invalid user placementRule for AVR %s", avr.Name)
+	}
 
-	for idx, subscription := range a.subscriptionList.Items {
-		// On the hub ignore any managed cluster subscriptions, as the hub maybe a managed cluster itself.
-		// SubscriptionSubscribed means this subscription is child sitting in managed cluster
-		// Placement.Local is true for a local subscription, and can be used in the absence of Status
-		if subscription.Status.Phase == subv1.SubscriptionSubscribed ||
-			(subscription.Spec.Placement != nil && subscription.Spec.Placement.Local != nil &&
-				*subscription.Spec.Placement.Local) {
-			a.log.Info("Skipping local subscription", "name", subscription.Name)
+	plRef := avr.Spec.Placement.PlacementRef
 
-			continue
-		}
+	if plRef.Namespace == "" {
+		plRef.Namespace = avr.Namespace
+	}
 
-		placementDecision, needRequeue := a.processSubscription(&a.subscriptionList.Items[idx])
+	userPlacementRule := &plrv1.PlacementRule{}
 
-		if needRequeue {
-			a.log.Info("Requeue for subscription", "name", subscription.Name)
+	err := r.Client.Get(ctx,
+		types.NamespacedName{Name: plRef.Name, Namespace: plRef.Namespace}, userPlacementRule)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get placementrule error: %w", err)
+	}
 
-			requeue = true
+	scName := userPlacementRule.Spec.SchedulerName
+	if scName != "" && scName != RamenScheduler {
+		return nil, fmt.Errorf("placementRule %s does not have the ramen scheduler. Scheduler used %s",
+			userPlacementRule.Name, scName)
+	}
 
-			continue
-		}
+	return userPlacementRule, nil
+}
 
-		if a.instance.Status.Decisions == nil {
-			a.instance.Status.Decisions = make(rmn.SubscriptionPlacementDecisionMap)
-		}
+func (r *ApplicationVolumeReplicationReconciler) annotatePlacementRule(ctx context.Context,
+	avr *rmn.ApplicationVolumeReplication, plRule *plrv1.PlacementRule) error {
+	if plRule.ObjectMeta.Annotations == nil {
+		plRule.ObjectMeta.Annotations = map[string]string{}
+	}
+
+	ownerName := plRule.ObjectMeta.Annotations[AVRNameAnnotation]
+	ownerNamespace := plRule.ObjectMeta.Annotations[AVRNamespaceAnnotation]
+
+	if ownerName == "" {
+		plRule.ObjectMeta.Annotations[AVRNameAnnotation] = avr.Name
+		plRule.ObjectMeta.Annotations[AVRNamespaceAnnotation] = avr.Namespace
 
-		if placementDecision != nil {
-			a.instance.Status.Decisions[subscription.Name] = placementDecision
+		err := r.Update(ctx, plRule)
```

Comment:
Tiny race window if 2 AVRs update one after the other having fetched an initial PlRule that was previously not owned.

Maybe there is a patch/apply option here that can be leveraged to apply a specific annotation patch if not present, to avoid the race.

Reply:
(alternate) instead of reaching the API server (as in, using apply or something, instead of update) and then failing, we could take local locks on userPlRule names, such that read/check/update of the annotation is within the lock, and the AVR reconciler hence goes one after the other when claiming ownership of a userPlRule in conflict across 2 or more AVRs.